### DISCUSSION TITLE
Difficulty Adjust Approach Rate in regards to Leaderboards/Score

### DISCUSSION THREAD

**Discussion by rozukke** [ppy/osu#20069] [Ideas]
As someone that uses osu!lazer as the default client to play osu!, Difficulty Adjust is the mod that probably has the highest number of use cases as well as the highest potential for upset and complaints. Personally I'm a fan of the customization potential it provides, but it comes with many issues in regards to leaderboards and scoring.

The categories of customization are CS, OD, HP and AR. Potential score multipliers for the first three are relatively self explanatory, and can be implemented with a simple vaguely linear curve, since all those options can be justified as "objective" difficulty. However, with AR the story is different, as everyone has a unique range which they find comfortable.

I've seen a worrying number of people suggest static points at which to increase and decrease score multiplier (such as >9.5, <6 etc.). This is worse than a band-aid solution, as it does not consider the context of the map whatsoever.

## Proposal
My personal proposal and the one I think many people will find logical is a system to prioritize mapper choice first and foremost, and operate on a **purely subtractive basis** originating from the maps preset AR.

![ar9graph](https://user-images.githubusercontent.com/73006620/187906042-303bbdb1-7e44-48ed-88e1-aa916daacd2d.png)
In the graph above, the AR of the map is 9, which corresponds to a 1x multiplier, and any AR increase or decrease is associated with a decrease in the score multiplier. This means that in terms of leaderboards, players who play the map closer to how it is intended to be played are rewarded with closer to a full score. 
This would mean that on an AR6 map, a player that changes it to be AR9.5 (much more readable for a majority of people) will likely not score higher on the leaderboard than someone that plays the map as intended.
This is very important for score and less so pp, since the system there accepts whatever value into difficulty calculation, while score needs to **consider the context of competitive play** on a map.

**All the values/curves are purely off the top of my head, as I want to portray the general idea more than a specific implementation.**

## Additional:
I've also heard discussions of mods like EZ having a positive score multiplier. Personally I am somewhat iffy on something like that since once again, the approach rate might be a worthwhile tradeoff for someone to use the CS and OD elements to make the map easier than it would be for them without the mod. However, in light of these discussions, I drafted up a modified version of the graph above that could implement a high and low shelf for AR ranges in the extremes for an AR7 map (sorry for the wonky graph, I am not a desmos wizard).

![ar7graph](https://user-images.githubusercontent.com/73006620/187906082-ba458819-d41f-4fce-b485-0e23ece82ddf.png)

This is also a very rough idea since AR is heavily associated with density, which I am not sure can be used for something like score multipliers readily.

*P.S. I am also aware that peppy wanted to avoid confusing score multipliers, but this is the best system I could think of* :)

**Comment by Flamiii** [comment]
I like the first graph/idea and I think that it makes a lot of sense, though I feel that for the sake of complexity you might just want to have it be on a linear scale instead of being exponential so that you end up with a simpler mod multiplier. So if a map is AR 9 like in your example, DA with AR <=8 or AR >=10 would be a 0.5x multiplier, AR 8.5 and AR 9.5 would be 0.75x, AR 9.4 would be 0.8x, AR 9.3 would be 0.85x, etc.

I know that you only touched on AR, but I feel that the same kind of idea for CS and OD could work as well. For example, if you're playing a CS 4 map with DA and CS 3 it would be a 0.5x multiplier, CS 3.5 would be 0.75x, CS 3.2 would be 0.6x, etc.

When it comes to OD, I would make it affect the mod multiplier by half as much since in my opinion it isn't as bad as decreasing CS by the same amount. For example, if you're playing an OD 9 map with DA and OD 8 it would be a 0.75x multiplier, OD 7.5 would be 0.625x, OD <=7 would be 0.5x, etc. 

And with something like CS or OD, increasing it above the map's initial value should just keep the mod multiplier at 1x (could potentially increase the mod multiplier but that belongs in its own separate discussion). I'm also not really sure what to do with HP but I think that's a topic for another day.

Combining all of these things together, if you're playing a CS 4 AR 9 OD 9 map with DA CS 3.5 AR 9.5 OD 7, the resulting mod multiplier would just be 0.75x * 0.75x  * 0.5x = 0.28125x. Some might say that this is still complicated but honestly I don't think it's really any more complex than EZHDHT (0.5x * 1.06x * 0.3x = 0.159x) or EZHTFLSO (0.5x * 0.3x * 1.12x * 0.9x = 0.1512x).

**Reply by thacer13** [reply]
For AR it's worth noting that the perceived change between, say, AR 5 and AR 5.5 is much smaller than AR 9 and AR 9.5, because of the percentage difference in milliseconds.

**Comment by Doksha** [comment]
an easier solution: any change in AR or OD would set the multiplier to 0.5; if CS or HP is increased, there is no multiplier applied. Your suggestion just seems way too complex and likely open to loopholes and just makes score setting so much weirder.

**Reply by rozukke** [reply]
This feels like a very extreme solution, since small adjustments realistically shouldn't chop resulting score in half (like adjustments of 0.1 AR). As for CS increases, it also does not feel fair if say CS4->CS7 gets treated as the same thing scorewise.

**Reply by Doksha** [reply]
tiny adjustments can be huge, and trying to have a sliding scale is just going to make things difficult. Increasing difficulty would be cool, but it would have to have a complex set of rules for difficulty increase to result in an increase of score. While a CS7 score *should* be worth more than a CS4 score, a CS5 score would not be worth the same difference. As peppy said, the user should be able to tell the score multiplier from a glance.
Basically, sliding scale is likely out of the question, so where do we draw the line on what counts as increase and what counts as a decrease

**Comment by peppy** [comment]
If a user cannot tell you what the resultant multiplier is for a mod being turned on, or a setting being adjusted, the implementation is flawed.

Score is supposed to be understandable. Keep this complex stuff for PP where it's a black box the user has no idea how it works.

**Reply by rozukke** [reply]
I would agree with this but the already existing multipliers feel quite arbitrary, and its also not exactly possible to intuitively feel out what a multiplier would be with a complex mod combo (something like 4 or more mods, especially the ones that aren't used very often). The infamous example of HD and HR having the same multiplier where most players disagree with this weighting shows that the multiplier system deserves tweaking and fine tuning to match more players' expectations. 

The alternative is to get rid of mod multipliers entirely, but the reactions to this depend heavily on each user (maybe people wouldn't like fragmented per-mod LBs etc.) and pp leaderboards would depend on a similarly unintuitive system that relies on pp being as objective as possible (which is very WIP). 

I don't see that much of a difference between score multipliers and pp in an ideal state, so I feel that at least with the mapper considerations that I have in mind for this suggestion, the score LBs would likely feel fairer with this sort of custom score multiplier idea (even if they're not as intuitive).

At the very least, I hope that what this brings up some ideas and considerations for the systems that are currently planned. Thank you for your time.

**Reply by thacer13** [reply]
Agreed, "If a user cannot tell you" sounds like odd criteria for something that may often add up to unguessable calculations. If you can see the final sum multiplier updating in real time, playing with settings could be made to feel intuitive imo.

**Reply by peppy** [reply]
I'm still against multipliers being a thing (but this is also up to community acceptance whether that can be a solution). They are going to become a second PP mess as per proposals like this.

We're not about to accept proposals like this into the code so please stay tuned as we get closer to merging scores and we can better figure a solution.
